python/GafferSceneUITest/SourceSetTest.py

```diff
@@ -0,0 +1,210 @@
+		s["a2"] = GafferScene.ShaderAssignment()
+		s["a2"]["in"].setInput( s["a"]["out"] )
+		s["a2"]["shader"].setInput( s["s2"]["out"] )
+
+		s["g2"] = GafferScene.Group()
+		s["g2"]["in"][0].setInput( s["a2"]["out"] )
+
+		s["o"] = GafferScene.SceneWriter()
+		s["o"]["in"].setInput( s["g2"]["out"] )
+
+		n = Gaffer.StandardSet()
+		c = Gaffer.Context()
+
+		self.assertEqual( len(GafferSceneUI.ContextAlgo.getSelectedPaths( c ).paths()), 0 )
+
+		h = GafferSceneUI.SourceSet( c, n )
+		self.assertEqual( h.size(), 0 )
+
+		a = "/group/group/planeA"
+		b = "/group/group/planeB"
+
+		n.add( s["o"] )
+
+		# Test scene selection changed
+
+		GafferSceneUI.ContextAlgo.setSelectedPaths( c, IECore.PathMatcher( [ a	 ] ) )
+		self.assertEqual( set( h ), set( [ s["pA"], ] ) )
+
+		GafferSceneUI.ContextAlgo.setSelectedPaths( c, IECore.PathMatcher( [ a, b ] ) )
+
+		self.assertEqual( set( h ), set( [ s["pA"], ] ) )
+
+		# Test defaulting to last input node if no valid plug or path
+		GafferSceneUI.ContextAlgo.setSelectedPaths( c, IECore.PathMatcher() )
+		self.assertEqual( set( h ), set( [ s["o"], ] ) )
+
+		# Test nodes changed
+
+		n.clear()
+		self.assertEqual( h.size(), 0 )
+
+		n.add( s["g2" ] )
+		GafferSceneUI.ContextAlgo.setSelectedPaths( c, IECore.PathMatcher( [ a ] ) )
+		self.assertEqual( set( h ), set( [ s["pA"], ] ) )
+
+		n.remove( s["g2"] )
+		n.add( s["g"] )
+		self.assertEqual( set( h ), set( [ s["g"], ] ) )
+
+		GafferSceneUI.ContextAlgo.setSelectedPaths( c, IECore.PathMatcher( [ "/group/planeA" ] ) )
+		self.assertEqual( set( h ), set( [ s["pA"], ] ) )
+
+		GafferSceneUI.ContextAlgo.setSelectedPaths( c, IECore.PathMatcher( [ "/group" ] ) )
+		n.clear()
+		n.add( s["g2"] )
+		self.assertEqual( set( h ), set( [ s["g2"], ] ) )
+
+		# Test incoming scene dirtied
+
+		GafferSceneUI.ContextAlgo.setSelectedPaths( c, IECore.PathMatcher( [ b ] ) )
+		self.assertEqual( set( h ), set( [ s["pB"], ] ) )
+
+		s["pB"]["name"].setValue( "notPlaneB" )
+		self.assertEqual( set( h ), set( [ s["g2"], ] ) )
+
+		# Test multiple non-scene nodes
+		n.clear()
+		n.add( s["s"] )
+		self.assertEqual( set( h ), set( [ s["s"], ] ) )
+		n.add( s["s2"] )
+		self.assertEqual( set( h ), set( [ s["s2"], ] ) )
+		n.remove( s["s2"] )
+		self.assertEqual( set( h ), set( [ s["s"], ] ) )
+
+	def testSignals( self ) :
+
+		s = Gaffer.ScriptNode()
+
+		s["pA"] = GafferScene.Plane()
+		s["pA"]["name"].setValue( "planeA" )
+		s["pB"] = GafferScene.Plane()
+		s["pB"]["name"].setValue( "planeB" )
+		s["g"] = GafferScene.Group()
+		s["g"]["in"][0].setInput( s["pA"]["out"] )
+		s["g"]["in"][1].setInput( s["pB"]["out"] )
+
+		s["g2"] = GafferScene.Group()
+		s["g2"]["in"][0].setInput( s["g"]["out"] )
+
+		mirror = set()
+
+		def added( _, member ) :
+			mirror.add( member )
+
+		def removed( _, member ) :
+			mirror.remove( member )
+
+		n = Gaffer.StandardSet()
+		c = Gaffer.Context()
+
+		h = GafferSceneUI.SourceSet( c, n )
+		ca = h.memberAddedSignal().connect( added )
```

Comment:
Not sure if you've come across it yet, or if you just don't think its that applicable here. But `GafferTest.CapturingSlot()` is very handy for testing signals - it simply accumulates a record of all the calls made to it.

Reply:
Ahh I hadn't seen that. It did seem like it's useful in this case to check the actual objects passed too and the resulting state accumulation? What do you think?

Reply:
I think you could argue that `CapturingSlot` would be better here. The current tests don't care what order the signals are emitted in, or how many times, just that they eventually agree with what we want. If we used CaptuingSlot, we'd be asserting that they were only triggered once per edit, and that remove was before add. Not a big deal though - I just wanted to make you aware of CapturingSlot, since it is frequently useful.